# LHL Capstone Project - Notebook #1 - EDA and Data Cleaning

In [1]:
# import packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# 1.0 - First Dataset - Hourly Generation from Jan 2023 - June 2023

## 1.1 - Initial EDA

In [2]:
# load first dataset
df_generation = pd.read_csv('CSD Generation (Hourly) - 2023-01 to 2023-06.csv')
df_generation.head()

Date (MST)           Date (MPT) Asset Short Name  \
0  2023-01-01 00:00:00  2023-01-01 00:00:00             AFG1   
1  2023-01-01 01:00:00  2023-01-01 01:00:00             AFG1   
2  2023-01-01 02:00:00  2023-01-01 02:00:00             AFG1   
3  2023-01-01 03:00:00  2023-01-01 03:00:00             AFG1   
4  2023-01-01 04:00:00  2023-01-01 04:00:00             AFG1   

           Asset Name Asset Grouping     Volume  Maximum Capability  \
0  AFG1 APF Athabasca           AFG1  82.586120               131.0   
1  AFG1 APF Athabasca           AFG1  81.336551               131.0   
2  AFG1 APF Athabasca           AFG1  81.498609               131.0   
3  AFG1 APF Athabasca           AFG1  81.882582               131.0   
4  AFG1 APF Athabasca           AFG1  82.032287               131.0   

   System Capability Fuel Type Sub Fuel Type  Planning Area     Region  
0               67.0     OTHER   Wood/Refuse             27  Northeast  
1               67.0     OTHER   Wood/Refuse             27  Northeast  
2               67.0     OTHER   Wood/Refuse             27  Northeast  
3               67.0     OTHER   Wood/Refuse             27  Northeast  
4               67.0     OTHER   Wood/Refuse             27  Northeast

In [3]:
df_generation.tail()

Date (MST)           Date (MPT) Asset Short Name  \
773280  2023-06-30 18:00:00  2023-06-30 19:00:00             WWD1   
773281  2023-06-30 19:00:00  2023-06-30 20:00:00             WWD1   
773282  2023-06-30 20:00:00  2023-06-30 21:00:00             WWD1   
773283  2023-06-30 21:00:00  2023-06-30 22:00:00             WWD1   
773284  2023-06-30 22:00:00  2023-06-30 23:00:00             WWD1   

           Asset Name Asset Grouping     Volume  Maximum Capability  \
773280  WWD1 Weldwood           WWD1  13.820833                50.0   
773281  WWD1 Weldwood           WWD1  14.089166                50.0   
773282  WWD1 Weldwood           WWD1  13.536333                50.0   
773283  WWD1 Weldwood           WWD1  12.905833                50.0   
773284  WWD1 Weldwood           WWD1  13.190833                50.0   

        System Capability Fuel Type Sub Fuel Type  Planning Area   Region  
773280               10.0     OTHER       Biomass             29  Central  
773281               10.0     OTHER       Biomass             29  Central  
773282               10.0     OTHER       Biomass             29  Central  
773283               10.0     OTHER       Biomass             29  Central  
773284               10.0     OTHER       Biomass             29  Central

In [4]:
df_generation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 773285 entries, 0 to 773284
Data columns (total 12 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Date (MST)          773285 non-null  object 
 1   Date (MPT)          773285 non-null  object 
 2   Asset Short Name    773285 non-null  object 
 3   Asset Name          773285 non-null  object 
 4   Asset Grouping      773285 non-null  object 
 5   Volume              773285 non-null  float64
 6   Maximum Capability  773285 non-null  float64
 7   System Capability   773285 non-null  float64
 8   Fuel Type           773285 non-null  object 
 9   Sub Fuel Type       773285 non-null  object 
 10  Planning Area       773285 non-null  int64  
 11  Region              773285 non-null  object 
dtypes: float64(3), int64(1), object(8)
memory usage: 70.8+ MB


In [5]:
# calculate the number of NaN values for volumes
df_generation.isnull().sum()

Date (MST)            0
Date (MPT)            0
Asset Short Name      0
Asset Name            0
Asset Grouping        0
Volume                0
Maximum Capability    0
System Capability     0
Fuel Type             0
Sub Fuel Type         0
Planning Area         0
Region                0
dtype: int64

In [6]:
# convert date values to datetime Dtype
df_generation['Date (MST)'] = pd.to_datetime(df_generation['Date (MST)'])
df_generation['Date (MPT)'] = pd.to_datetime(df_generation['Date (MPT)'])

In [7]:
# confirm dtype change
df_generation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 773285 entries, 0 to 773284
Data columns (total 12 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   Date (MST)          773285 non-null  datetime64[ns]
 1   Date (MPT)          773285 non-null  datetime64[ns]
 2   Asset Short Name    773285 non-null  object        
 3   Asset Name          773285 non-null  object        
 4   Asset Grouping      773285 non-null  object        
 5   Volume              773285 non-null  float64       
 6   Maximum Capability  773285 non-null  float64       
 7   System Capability   773285 non-null  float64       
 8   Fuel Type           773285 non-null  object        
 9   Sub Fuel Type       773285 non-null  object        
 10  Planning Area       773285 non-null  int64         
 11  Region              773285 non-null  object        
dtypes: datetime64[ns](2), float64(3), int64(1), object(6)
memory usage: 70.8+ MB


In [8]:
df_generation.describe()

Date (MST)                     Date (MPT)  \
count                         773285                         773285   
mean   2023-04-01 21:58:34.148600576  2023-04-01 22:35:30.274866432   
min              2023-01-01 00:00:00            2023-01-01 00:00:00   
25%              2023-02-15 14:00:00            2023-02-15 14:00:00   
50%              2023-04-02 02:00:00            2023-04-02 03:00:00   
75%              2023-05-17 08:00:00            2023-05-17 09:00:00   
max              2023-06-30 22:00:00            2023-06-30 23:00:00   
std                              NaN                            NaN   

              Volume  Maximum Capability  System Capability  Planning Area  
count  773285.000000       773285.000000      773285.000000  773285.000000  
mean       54.493146          103.404604         100.961400      35.635644  
min       -25.159333            5.000000           0.000000       4.000000  
25%         0.000000           19.000000          16.000000      25.000000  
50%         9.586333           48.000000          48.000000      39.000000  
75%        55.610177          120.000000         120.000000      52.000000  
max       856.374077          868.000000         868.000000      60.000000  
std       105.359007          135.654632         136.202336      16.318727

In [9]:
# add new column "asset utilization"
df_generation['Asset Utilization'] = df_generation['Volume'] / df_generation['Maximum Capability']

## 1.2 - Assets, fuel type and regions EDA

In [10]:
# How many unique assets are in this dataset?
generation_assets = df_generation['Asset Short Name'].value_counts().index.tolist()
print(len(generation_assets))

183


In [11]:
# What are the different fuel type categories?
fuel_types = df_generation['Fuel Type'].value_counts().index.tolist()
fuel_types

['GAS',
 'WIND',
 'SOLAR',
 'OTHER',
 'HYDRO',
 'ENERGY STORAGE',
 'COAL',
 'DUAL FUEL']

In [12]:
# How many different regions are there?
regions = df_generation['Region'].value_counts().index.tolist()
regions

['South', 'Central', 'Northwest', 'Northeast', 'Edmonton', 'Calgary']

#### There are six different regions that assets are classified into.

In [13]:
# How many assets are in each region?
assets_per_region = df_generation.copy().drop_duplicates(subset=['Asset Short Name'])
assets_per_region['Region'].value_counts()

Region
South        80
Central      32
Northwest    28
Northeast    19
Edmonton     15
Calgary       9
Name: count, dtype: int64

#### The majority of assets are in the South region. 

In [14]:
# Let's check to see what the breakdown of fuel types in the South region is
south_region = assets_per_region[assets_per_region['Region'] == 'South']
south_region['Fuel Type'].value_counts()

Fuel Type
WIND              29
SOLAR             27
GAS               16
HYDRO              6
OTHER              1
ENERGY STORAGE     1
Name: count, dtype: int64

#### Most energy assets in the South region are clean tech (Wind and Solar). Knowing that the oil sands area is in Northern Alberta, this makes sense, but let's confirm by repeating this check for the North regions.

In [15]:
north_region = assets_per_region[(assets_per_region['Region'] == 'Northwest') | (assets_per_region['Region'] == 'Northeast')]
north_region['Fuel Type'].value_counts()

Fuel Type
GAS               35
OTHER             10
ENERGY STORAGE     2
Name: count, dtype: int64

In [16]:
# How many assets are in each fuel category?
assets_per_region['Fuel Type'].value_counts()

Fuel Type
GAS               82
WIND              37
SOLAR             34
OTHER             13
HYDRO              9
ENERGY STORAGE     5
COAL               2
DUAL FUEL          1
Name: count, dtype: int64

In [17]:
# What is the average system capability of each fuel type?
(assets_per_region.groupby('Fuel Type')['System Capability'].sum())/(assets_per_region.groupby('Fuel Type')['Asset Name'].count())

Fuel Type
COAL              410.000000
DUAL FUEL         466.000000
ENERGY STORAGE     18.000000
GAS               129.123171
HYDRO              99.366667
OTHER              25.284615
SOLAR              36.587647
WIND              100.837838
dtype: float64

In [18]:
# What is the total capacity of each region?
assets_per_region.groupby('Region')['System Capability'].sum().sort_values(ascending=False)

Region
South        5878.00
Northeast    3622.00
Central      3123.70
Edmonton     2833.98
Calgary      1505.80
Northwest    1198.60
Name: System Capability, dtype: float64

#### The South region has the highest total system capability, which is correlated with having the most assets. Now, let's check to see what the average system capability is when adjusted for the number of assets in that region to determine the energy density of each region.

In [19]:
(assets_per_region.groupby('Region')['System Capability'].sum()/(assets_per_region.groupby('Region')['Fuel Type'].count())).sort_values(ascending=False)

Region
Northeast    190.631579
Edmonton     188.932000
Calgary      167.311111
Central       97.615625
South         73.475000
Northwest     42.807143
dtype: float64

#### When we analyze the regions by energy density, however, the South region has the second lowest energy density, almost 3x lower than the Northeast region. This can be attributed to clean energy assets having lower energy densities than natural gas.

In [20]:
# Total system capability volume for each fuel type
assets_per_region.groupby('Fuel Type')['System Capability'].sum().sort_values(ascending=False)

Fuel Type
GAS               10588.10
WIND               3731.00
SOLAR              1243.98
HYDRO               894.30
COAL                820.00
DUAL FUEL           466.00
OTHER               328.70
ENERGY STORAGE       90.00
Name: System Capability, dtype: float64

## 1.3 - Time-Series EDA

In [21]:
# Create pivot table of each fuel type's volume at each hour.
hourly_volume_df = pd.pivot_table(df_generation, index = ['Date (MST)'], values = 'Volume', columns=['Fuel Type'], aggfunc='sum')
hourly_volume_df

Fuel Type                  COAL   DUAL FUEL  ENERGY STORAGE          GAS  \
Date (MST)                                                                 
2023-01-01 00:00:00  817.176320  464.670162        0.000000  7598.055380   
2023-01-01 01:00:00  817.141767  464.099606        0.000000  7669.070478   
2023-01-01 02:00:00  816.632098  464.155770        0.000000  7578.895878   
2023-01-01 03:00:00  817.109768  464.372050        0.000000  7518.248410   
2023-01-01 04:00:00  816.817543  463.750606        0.000000  7536.114298   
...                         ...         ...             ...          ...   
2023-06-30 18:00:00  791.614155  465.974161        0.047222  7123.095844   
2023-06-30 19:00:00  790.207208  465.677050       -0.004389  7156.218349   
2023-06-30 20:00:00  792.104988  465.763883       -0.003778  7010.609269   
2023-06-30 21:00:00  793.254376  465.957605       -0.000222  6848.686789   
2023-06-30 22:00:00  793.422655  466.047328        0.008667  6621.581757   

Fuel Type                 HYDRO       OTHER       SOLAR        WIND  
Date (MST)                                                           
2023-01-01 00:00:00  116.953446  287.754563    0.000000  780.206753  
2023-01-01 01:00:00  114.114128  283.552572    0.000000  732.207446  
2023-01-01 02:00:00  121.666094  281.975430    0.000000  743.884060  
2023-01-01 03:00:00  113.459167  283.418738    0.000000  759.293766  
2023-01-01 04:00:00  139.900810  282.133304    0.000000  746.124896  
...                         ...         ...         ...         ...  
2023-06-30 18:00:00  624.109243  280.019205  444.602942  459.752502  
2023-06-30 19:00:00  614.766657  290.221684  204.177766  446.869762  
2023-06-30 20:00:00  603.492175  274.901387   13.199218  640.186719  
2023-06-30 21:00:00  592.913197  264.828780    0.000000  824.856876  
2023-06-30 22:00:00  576.197029  267.370789    0.000000  934.051413  

[4343 rows x 8 columns]

In [22]:
# Create new column that adds up the energy produced each hour
hourly_volume_df['Total Volume (MW)'] = hourly_volume_df[fuel_types].sum(axis=1)
hourly_volume_df

Fuel Type                  COAL   DUAL FUEL  ENERGY STORAGE          GAS  \
Date (MST)                                                                 
2023-01-01 00:00:00  817.176320  464.670162        0.000000  7598.055380   
2023-01-01 01:00:00  817.141767  464.099606        0.000000  7669.070478   
2023-01-01 02:00:00  816.632098  464.155770        0.000000  7578.895878   
2023-01-01 03:00:00  817.109768  464.372050        0.000000  7518.248410   
2023-01-01 04:00:00  816.817543  463.750606        0.000000  7536.114298   
...                         ...         ...             ...          ...   
2023-06-30 18:00:00  791.614155  465.974161        0.047222  7123.095844   
2023-06-30 19:00:00  790.207208  465.677050       -0.004389  7156.218349   
2023-06-30 20:00:00  792.104988  465.763883       -0.003778  7010.609269   
2023-06-30 21:00:00  793.254376  465.957605       -0.000222  6848.686789   
2023-06-30 22:00:00  793.422655  466.047328        0.008667  6621.581757   

Fuel Type                 HYDRO       OTHER       SOLAR        WIND  \
Date (MST)                                                            
2023-01-01 00:00:00  116.953446  287.754563    0.000000  780.206753   
2023-01-01 01:00:00  114.114128  283.552572    0.000000  732.207446   
2023-01-01 02:00:00  121.666094  281.975430    0.000000  743.884060   
2023-01-01 03:00:00  113.459167  283.418738    0.000000  759.293766   
2023-01-01 04:00:00  139.900810  282.133304    0.000000  746.124896   
...                         ...         ...         ...         ...   
2023-06-30 18:00:00  624.109243  280.019205  444.602942  459.752502   
2023-06-30 19:00:00  614.766657  290.221684  204.177766  446.869762   
2023-06-30 20:00:00  603.492175  274.901387   13.199218  640.186719   
2023-06-30 21:00:00  592.913197  264.828780    0.000000  824.856876   
2023-06-30 22:00:00  576.197029  267.370789    0.000000  934.051413   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-01-01 00:00:00       10064.816624  
2023-01-01 01:00:00       10080.185996  
2023-01-01 02:00:00       10007.209331  
2023-01-01 03:00:00        9955.901898  
2023-01-01 04:00:00        9984.841457  
...                                ...  
2023-06-30 18:00:00       10189.215274  
2023-06-30 19:00:00        9968.134088  
2023-06-30 20:00:00        9800.253862  
2023-06-30 21:00:00        9790.497400  
2023-06-30 22:00:00        9658.679638  

[4343 rows x 9 columns]

In [23]:
# Find the top 5 hours with the highest Total Volume
hourly_volume_df.sort_values('Total Volume (MW)', ascending=False)

Fuel Type                   COAL   DUAL FUEL  ENERGY STORAGE           GAS  \
Date (MST)                                                                   
2023-06-25 22:00:00  1598.326415  898.748988        0.011444  12515.340401   
2023-01-23 10:00:00   817.470098  465.572827       45.741942   7943.036332   
2023-01-04 11:00:00   814.401373  463.833160        0.000000   7677.549624   
2023-02-24 14:00:00   814.341375  463.992550       -5.072503   7884.296063   
2023-02-24 15:00:00   815.415099  463.379159       -3.862059   7773.999193   
...                          ...         ...             ...           ...   
2023-05-07 01:00:00   302.724660  465.472104       -0.990444   5677.246546   
2023-05-28 05:00:00   803.744156  464.873881       -0.995889   5252.110418   
2023-05-28 04:00:00   803.614267  464.904773       -2.725722   5331.855850   
2023-05-07 02:00:00   298.398160  465.296827        0.000000   5629.051228   
2023-05-07 03:00:00   296.866327  465.299994        0.000000   5658.824619   

Fuel Type                  HYDRO       OTHER       SOLAR         WIND  \
Date (MST)                                                              
2023-06-25 22:00:00  1175.068039  515.147144    0.000000  1447.147752   
2023-01-23 10:00:00   140.702962  271.054301   51.532166  1923.733835   
2023-01-04 11:00:00   162.139227  269.446785  567.215049  1695.409923   
2023-02-24 14:00:00    87.971885  261.618407  628.555724  1509.295260   
2023-02-24 15:00:00   106.894894  265.549528  649.920766  1542.132724   
...                          ...         ...         ...          ...   
2023-05-07 01:00:00   156.819058  209.474652    0.000000   621.639687   
2023-05-28 05:00:00   182.629757  256.220025  102.951032   361.649010   
2023-05-28 04:00:00   201.649363  254.425480    5.092000   356.830461   
2023-05-07 02:00:00   175.325419  208.868525    0.000000   632.341812   
2023-05-07 03:00:00   198.523500  209.025106    0.000000   567.860398   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-06-25 22:00:00       18149.790184  
2023-01-23 10:00:00       11658.844463  
2023-01-04 11:00:00       11649.995142  
2023-02-24 14:00:00       11644.998762  
2023-02-24 15:00:00       11613.429304  
...                                ...  
2023-05-07 01:00:00        7432.386263  
2023-05-28 05:00:00        7423.182391  
2023-05-28 04:00:00        7415.646471  
2023-05-07 02:00:00        7409.281971  
2023-05-07 03:00:00        7396.399944  

[4343 rows x 9 columns]

#### Something looks off with the volumes for June 25 at 22:00. Total volume is nearly double the next highest value, and there is no value from a date or time close as there is for Feb 24. Sort each column to see if these values are reasonable.

In [24]:
# COAL
hourly_volume_df.sort_values('COAL', ascending=False).head()

Fuel Type                   COAL   DUAL FUEL  ENERGY STORAGE           GAS  \
Date (MST)                                                                   
2023-06-25 22:00:00  1598.326415  898.748988        0.011444  12515.340401   
2023-05-11 07:00:00   821.208544  465.905552       -3.273611   6490.733076   
2023-01-26 03:00:00   821.047543  465.288329      -16.060499   6008.801801   
2023-01-25 18:00:00   821.004379  465.615605      -18.468000   7729.617693   
2023-05-10 12:00:00   820.982321  465.597827        0.000000   6656.412697   

Fuel Type                  HYDRO       OTHER       SOLAR         WIND  \
Date (MST)                                                              
2023-06-25 22:00:00  1175.068039  515.147144    0.000000  1447.147752   
2023-05-11 07:00:00   248.604808  197.558225  440.170368    83.239467   
2023-01-26 03:00:00    96.440690  244.448545    0.000000  2421.761178   
2023-01-25 18:00:00   131.041619  296.065357    0.000000   952.982326   
2023-05-10 12:00:00   329.175692  218.450641  500.765348    15.849817   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-06-25 22:00:00       18149.790184  
2023-05-11 07:00:00        8744.146428  
2023-01-26 03:00:00       10041.727586  
2023-01-25 18:00:00       10377.858979  
2023-05-10 12:00:00        9007.234344

#### The total maximum capability of all coal assets is 820 MW; this value is not reasonable and will need manual adjustment.

In [25]:
# DUAL FUEL
hourly_volume_df.sort_values('DUAL FUEL', ascending=False).head()

Fuel Type                   COAL   DUAL FUEL  ENERGY STORAGE           GAS  \
Date (MST)                                                                   
2023-06-25 22:00:00  1598.326415  898.748988        0.011444  12515.340401   
2023-01-28 01:00:00   816.530265  468.579993       -1.304999   7868.780348   
2023-01-28 02:00:00   817.077546  468.225492       -1.344222   7829.178474   
2023-01-28 03:00:00   816.649376  468.182772       -0.826222   7615.703411   
2023-01-28 06:00:00   817.006212  468.159104        0.000000   8130.708701   

Fuel Type                  HYDRO       OTHER  SOLAR         WIND  \
Date (MST)                                                         
2023-06-25 22:00:00  1175.068039  515.147144    0.0  1447.147752   
2023-01-28 01:00:00   102.269185  269.669179    0.0   655.614772   
2023-01-28 02:00:00    99.999995  262.881059    0.0   600.249164   
2023-01-28 03:00:00   101.722518  261.514524    0.0   510.290868   
2023-01-28 06:00:00    98.819896  298.026094    0.0   255.407237   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-06-25 22:00:00       18149.790184  
2023-01-28 01:00:00       10180.138742  
2023-01-28 02:00:00       10076.267509  
2023-01-28 03:00:00        9773.237247  
2023-01-28 06:00:00       10068.127245

#### The total maximum capability of all dual fuel assets is 466 MW; this value is not reasonable and will need manual adjustment.

In [26]:
# ENERGY STORAGE
hourly_volume_df.sort_values('ENERGY STORAGE', ascending=False).head()

Fuel Type                  COAL   DUAL FUEL  ENERGY STORAGE          GAS  \
Date (MST)                                                                 
2023-01-25 12:00:00  390.398717  466.200493       74.199997  7819.634542   
2023-01-25 11:00:00  349.881326  465.180105       74.199997  7877.586017   
2023-01-25 10:00:00  343.651382  465.879770       73.067610  7944.203226   
2023-01-22 16:00:00  798.537489  465.961272       70.400002  7664.637819   
2023-01-25 09:00:00  348.703217  465.825216       67.956111  7874.940887   

Fuel Type                 HYDRO       OTHER       SOLAR         WIND  \
Date (MST)                                                             
2023-01-25 12:00:00  150.087012  302.808290  193.421778   916.353324   
2023-01-25 11:00:00  156.701770  308.924768  115.745584   944.257871   
2023-01-25 10:00:00  199.644974  307.153242   73.424654   935.896946   
2023-01-22 16:00:00  151.940293  264.805348   71.594813  1105.794401   
2023-01-25 09:00:00  155.540206  304.418734   64.469983  1137.308376   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-01-25 12:00:00       10313.104154  
2023-01-25 11:00:00       10292.477438  
2023-01-25 10:00:00       10342.921805  
2023-01-22 16:00:00       10593.671436  
2023-01-25 09:00:00       10419.162729

#### The total maximum capability of all energy storage assets is 90 MW; this value is reasonable.

In [27]:
# GAS
hourly_volume_df.sort_values('GAS', ascending=False).head()

Fuel Type                   COAL   DUAL FUEL  ENERGY STORAGE           GAS  \
Date (MST)                                                                   
2023-06-25 22:00:00  1598.326415  898.748988        0.011444  12515.340401   
2023-02-23 19:00:00   816.633208  463.152328        0.000000   8811.276138   
2023-02-23 18:00:00   815.158988  464.353048        0.000000   8778.727813   
2023-02-23 20:00:00   816.312156  463.303828        0.000000   8716.771437   
2023-02-23 17:00:00   816.007767  464.542938        0.000000   8665.363268   

Fuel Type                  HYDRO       OTHER      SOLAR         WIND  \
Date (MST)                                                             
2023-06-25 22:00:00  1175.068039  515.147144   0.000000  1447.147752   
2023-02-23 19:00:00   218.302812  297.895176   4.171444    13.605905   
2023-02-23 18:00:00   268.059066  300.198068   9.749055    15.446089   
2023-02-23 20:00:00   220.174008  300.599188   0.000000    22.033477   
2023-02-23 17:00:00   282.821984  292.381752  49.301388    13.658072   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-06-25 22:00:00       18149.790184  
2023-02-23 19:00:00       10625.037013  
2023-02-23 18:00:00       10651.692127  
2023-02-23 20:00:00       10539.194094  
2023-02-23 17:00:00       10584.077169

#### The total maximum capability of all gas assets is 10,904 MW, this value is not reasonable and will need manual adjustment.

In [28]:
# HYDRO
hourly_volume_df.sort_values('HYDRO', ascending=False).head()

Fuel Type                   COAL   DUAL FUEL  ENERGY STORAGE           GAS  \
Date (MST)                                                                   
2023-06-25 22:00:00  1598.326415  898.748988        0.011444  12515.340401   
2023-06-26 15:00:00   797.074988  462.761938        0.002444   7304.069958   
2023-06-26 16:00:00   794.573542  462.573105        0.002167   7408.687609   
2023-06-26 14:00:00   797.653655  463.126106        0.000667   7062.926390   
2023-06-23 18:00:00   798.056154  464.547828       -0.845000   6553.472213   

Fuel Type                  HYDRO       OTHER       SOLAR         WIND  \
Date (MST)                                                              
2023-06-25 22:00:00  1175.068039  515.147144    0.000000  1447.147752   
2023-06-26 15:00:00   690.780322  241.054960  939.765948    53.404075   
2023-06-26 16:00:00   689.732919  240.218852  858.245788    23.863807   
2023-06-26 14:00:00   662.033012  242.102341  999.045361    42.936763   
2023-06-23 18:00:00   656.645060  259.778286  641.026582   307.076796   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-06-25 22:00:00       18149.790184  
2023-06-26 15:00:00       10488.914634  
2023-06-26 16:00:00       10477.897789  
2023-06-26 14:00:00       10269.824295  
2023-06-23 18:00:00        9679.757918

#### The total maximum capability of all hydro assets is 894 MW; this value is not reasonable and will need manual adjustment.

In [29]:
# OTHER
hourly_volume_df.sort_values('OTHER', ascending=False).head()

Fuel Type                   COAL   DUAL FUEL  ENERGY STORAGE           GAS  \
Date (MST)                                                                   
2023-06-25 22:00:00  1598.326415  898.748988        0.011444  12515.340401   
2023-03-27 20:00:00   812.560209  465.008882        0.000000   7772.500943   
2023-03-27 21:00:00   785.920821  465.161715        0.000000   7548.564661   
2023-03-27 19:00:00   813.458210  465.147773        0.000000   7864.407946   
2023-03-27 04:00:00   764.037155  465.578770        0.000000   7287.857545   

Fuel Type                  HYDRO       OTHER     SOLAR         WIND  \
Date (MST)                                                            
2023-06-25 22:00:00  1175.068039  515.147144  0.000000  1447.147752   
2023-03-27 20:00:00   128.921608  336.669533  0.408909    70.986775   
2023-03-27 21:00:00   116.604531  336.422619  0.048615    93.520962   
2023-03-27 19:00:00   179.552839  336.379700  0.408909    48.667132   
2023-03-27 04:00:00   135.726588  335.282567  0.000000    49.950620   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-06-25 22:00:00       18149.790184  
2023-03-27 20:00:00        9587.056859  
2023-03-27 21:00:00        9346.243924  
2023-03-27 19:00:00        9708.022508  
2023-03-27 04:00:00        9038.433244

#### The total maximum capability of all "other" assets is 444 MW; this value is not reasonable and will need manual adjustment.

In [30]:
# SOLAR
hourly_volume_df.sort_values('SOLAR', ascending=False).head()

Fuel Type                  COAL   DUAL FUEL  ENERGY STORAGE          GAS  \
Date (MST)                                                                 
2023-05-30 10:00:00  798.457044  464.780550        0.010500  5671.680630   
2023-06-22 10:00:00  802.065154  464.589883        0.008000  6349.154640   
2023-06-04 13:00:00  804.038546  464.483940        0.009167  6258.715052   
2023-06-04 12:00:00  796.570547  464.487829       -0.297278  6050.922378   
2023-05-30 11:00:00  798.896208  464.969273        0.001556  5812.842368   

Fuel Type                 HYDRO       OTHER        SOLAR        WIND  \
Date (MST)                                                             
2023-05-30 10:00:00  360.266669  262.938279  1127.707101  627.864117   
2023-06-22 10:00:00  651.453552  226.608741  1117.544015  187.810847   
2023-06-04 13:00:00  277.215517  254.818190  1112.223044  395.727956   
2023-06-04 12:00:00  239.351391  245.644754  1111.584455  418.615401   
2023-05-30 11:00:00  367.863569  269.225679  1108.434730  603.506567   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-05-30 10:00:00        9313.704890  
2023-06-22 10:00:00        9799.234833  
2023-06-04 13:00:00        9567.231413  
2023-06-04 12:00:00        9326.879477  
2023-05-30 11:00:00        9425.739951

#### The total maximum capability of all solar assets is 1,248 MW; this value is reasonable.

In [31]:
# WIND
hourly_volume_df.sort_values('WIND', ascending=False).head()

Fuel Type                  COAL   DUAL FUEL  ENERGY STORAGE          GAS  \
Date (MST)                                                                 
2023-04-08 20:00:00  639.750712  465.403327        0.000000  5610.691032   
2023-04-09 01:00:00  406.663327  465.292215        0.000000  5390.680817   
2023-04-08 23:00:00  508.994213  465.875938        0.000000  5348.342805   
2023-04-08 21:00:00  608.467435  465.445438        0.024333  5576.745789   
2023-04-09 00:00:00  424.674493  465.724716        0.000000  5449.491387   

Fuel Type                 HYDRO       OTHER  SOLAR         WIND  \
Date (MST)                                                        
2023-04-08 20:00:00  130.583712  257.533953    0.0  2824.874941   
2023-04-09 01:00:00  104.942698  254.659780    0.0  2806.504006   
2023-04-08 23:00:00  119.018572  255.719340    0.0  2795.389370   
2023-04-08 21:00:00  185.809836  255.019822    0.0  2784.737182   
2023-04-09 00:00:00  115.200932  254.961407    0.0  2772.611408   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-04-08 20:00:00        9928.837676  
2023-04-09 01:00:00        9428.742842  
2023-04-08 23:00:00        9493.340238  
2023-04-08 21:00:00        9876.249837  
2023-04-09 00:00:00        9482.664344

#### The total maximum capability of all wind assets is 3,731 MW; this value is reasonable.

**Values for June 25 at 22:00 need to be adjusted for other, hydro, gas, dual fuel, and coal fuel types.**

In [32]:
# convert the index to a datetime column dtype to filter for the values that need adjustment
hourly_volume_df.index = pd.to_datetime(hourly_volume_df.index)
hourly_volume_df.loc[(hourly_volume_df.index >= '2023-06-24 22:00') & (hourly_volume_df.index <= '2023-06-26 22:00')]

Fuel Type                   COAL   DUAL FUEL  ENERGY STORAGE           GAS  \
Date (MST)                                                                   
2023-06-24 22:00:00   794.052874  461.520771        0.000000   5739.874455   
2023-06-24 23:00:00   793.015323  449.293716        0.000611   5389.778058   
2023-06-25 00:00:00   708.526540  461.702327       -0.295778   5434.924425   
2023-06-25 01:00:00   624.525879  463.442884       -3.746110   5600.741767   
2023-06-25 02:00:00   603.889938  463.583328       -3.734221   5594.726871   
2023-06-25 03:00:00   604.015101  464.037883       -1.661222   5615.628955   
2023-06-25 04:00:00   603.888159  464.355049        0.001833   5772.136668   
2023-06-25 05:00:00   681.687878  464.976382        0.005611   5647.599308   
2023-06-25 06:00:00   782.372931  465.089159       -0.000333   5578.257480   
2023-06-25 07:00:00   799.420766  464.924883        0.009611   5708.006910   
2023-06-25 08:00:00   800.229321  464.907269       -0.226333   5864.533676   
2023-06-25 09:00:00   800.582821  464.908827       -1.948333   5968.299202   
2023-06-25 10:00:00   800.119269  464.809994       -0.496555   6045.980220   
2023-06-25 11:00:00   800.463599  464.849050        0.004611   6082.903938   
2023-06-25 12:00:00   794.664209  464.588717        0.002000   6113.025453   
2023-06-25 13:00:00   796.344766  464.361660        0.000056   6157.737378   
2023-06-25 14:00:00   797.461988  464.153496        0.002056   6304.405600   
2023-06-25 15:00:00   795.858654  464.148217        0.003667   6442.605832   
2023-06-25 16:00:00   797.704379  464.070049        0.008389   6376.486477   
2023-06-25 17:00:00   798.205878  463.984329        0.004278   6608.320016   
2023-06-25 18:00:00   798.606655  464.274884        0.021833   6671.958190   
2023-06-25 19:00:00   798.536266  464.684106        0.001222   6658.370183   
2023-06-25 20:00:00   798.378210  464.401160       -0.000667   6605.446860   
2023-06-25 21:00:00   798.246821  463.050217       -0.000611   6567.603357   
2023-06-25 22:00:00  1598.326415  898.748988        0.011444  12515.340401   
2023-06-25 23:00:00   798.398932  463.328661       -1.033500   6244.736352   
2023-06-26 00:00:00   798.903932  464.653049       -0.601389   6243.219483   
2023-06-26 01:00:00   799.590321  464.866772       -2.786555   6168.344914   
2023-06-26 02:00:00   797.826266  464.826994       -2.805888   6334.097676   
2023-06-26 03:00:00   798.301433  464.782715       -2.803833   6338.304135   
2023-06-26 04:00:00   798.471601  464.730217       -0.383444   6218.583303   
2023-06-26 05:00:00   798.480265  464.702495        0.000611   6427.865464   
2023-06-26 06:00:00   798.660880  464.569104        0.001444   6533.236937   
2023-06-26 07:00:00   798.972267  464.146772       10.111000   6665.533142   
2023-06-26 08:00:00   797.055320  464.011272        0.003056   6802.691227   
2023-06-26 09:00:00   797.456099  463.954272        0.001778   6653.896067   
2023-06-26 10:00:00   798.229320  463.825104        0.005500   6832.519966   
2023-06-26 11:00:00   798.341878  464.073050        0.018278   6990.306278   
2023-06-26 12:00:00   798.052155  463.753660        0.022722   7031.377080   
2023-06-26 13:00:00   798.012377  463.518549        0.008500   7069.834385   
2023-06-26 14:00:00   797.653655  463.126106        0.000667   7062.926390   
2023-06-26 15:00:00   797.074988  462.761938        0.002444   7304.069958   
2023-06-26 16:00:00   794.573542  462.573105        0.002167   7408.687609   
2023-06-26 17:00:00   794.008375  462.426271        0.002500   7158.818827   
2023-06-26 18:00:00   793.912431  462.143215        0.004000   7159.499244   
2023-06-26 19:00:00   793.725708  462.111883        0.015167   7348.623403   
2023-06-26 20:00:00   794.570656  462.328495       -0.007222   7301.567534   
2023-06-26 21:00:00   793.746098  462.860772        0.000667   7123.653536   
2023-06-26 22:00:00   794.865376  463.426383       -0.001778   6745.604461   

Fuel Type            

In [33]:
# Go back to first dataframe and observe all data for the time in question for outliers.
june_25 = df_generation.loc[df_generation['Date (MST)'] == '2023-06-25 22:00']
june_25.sort_values('Volume', ascending = False)

Date (MST)          Date (MPT) Asset Short Name  \
420227 2023-06-25 22:00:00 2023-06-25 23:00:00             EGC1   
758499 2023-06-25 22:00:00 2023-06-25 23:00:00             EGC1   
760677 2023-06-25 22:00:00 2023-06-25 23:00:00              GN3   
468066 2023-06-25 22:00:00 2023-06-25 23:00:00              GN3   
763944 2023-06-25 22:00:00 2023-06-25 23:00:00              KH3   
...                    ...                 ...              ...   
754990 2023-06-25 22:00:00 2023-06-25 23:00:00             CHP1   
347245 2023-06-25 22:00:00 2023-06-25 23:00:00             CLR2   
437189 2023-06-25 22:00:00 2023-06-25 23:00:00             ERV2   
755353 2023-06-25 22:00:00 2023-06-25 23:00:00             CLR2   
773164 2023-06-25 22:00:00 2023-06-25 23:00:00             WWD1   

                          Asset Name Asset Grouping      Volume  \
420227                  EGC1 Shepard           EGC1  640.425430   
758499                  EGC1 Shepard           EGC1  640.425430   
760677                GN3 Genesee #3            GN3  449.374494   
468066                GN3 Genesee #3            GN3  449.374494   
763944              KH3 Keephills #3            KH3  421.554772   
...                              ...            ...         ...   
754990            CHP1 Chappice Lake           CHP1    0.000000   
347245               CLR2 Claresholm           CLR2    0.000000   
437189  ERV2 eReserve2 Buffalo Creek           ERV2    0.000000   
755353               CLR2 Claresholm           CLR2    0.000000   
773164                 WWD1 Weldwood           WWD1    0.000000   

        Maximum Capability  System Capability       Fuel Type  \
420227               868.0              868.0             GAS   
758499               868.0              868.0             GAS   
760677               466.0              466.0       DUAL FUEL   
468066               466.0              466.0       DUAL FUEL   
763944               463.0              463.0             GAS   
...                    ...                ...             ...   
754990                14.0               14.0           SOLAR   
347245                75.0               75.0           SOLAR   
437189                20.0               20.0  ENERGY STORAGE   
755353                75.0               75.0           SOLAR   
773164                50.0               10.0           OTHER   

          Sub Fuel Type  Planning Area    Region  Asset Utilization  
420227   COMBINED_CYCLE              6   Calgary           0.737817  
758499   COMBINED_CYCLE              6   Calgary           0.737817  
760677        DUAL FUEL             40  Edmonton           0.964323  
468066        DUAL FUEL             40  Edmonton           0.964323  
763944  GAS_FIRED_STEAM             40  Edmonton           0.910485  
...                 ...            ...       ...                ...  
754990            SOLAR              4     South           0.000000  
347245            SOLAR             49     South           0.000000  
437189   ENERGY STORAGE             32   Central           0.000000  
755353            SOLAR             49     South           0.000000  
773164          Biomass             29   Central           0.000000  

[364 rows x 13 columns]

#### There are 183 assets in this dataset, yet the above data frame shows 364 records for 1 hour, and we can see that there are duplicated rows. Check how many rows are returned for June 24 and 26

In [34]:
# June 24
june_24 = df_generation.loc[df_generation['Date (MST)'] == '2023-06-24 22:00']
june_24.sort_values('Volume', ascending = False)

Date (MST)          Date (MPT) Asset Short Name  \
420203 2023-06-24 22:00:00 2023-06-24 23:00:00             EGC1   
468042 2023-06-24 22:00:00 2023-06-24 23:00:00              GN3   
541144 2023-06-24 22:00:00 2023-06-24 23:00:00              KH3   
637926 2023-06-24 22:00:00 2023-06-24 23:00:00             SCL1   
462428 2023-06-24 22:00:00 2023-06-24 23:00:00              GN1   
...                    ...                 ...              ...   
451200 2023-06-24 22:00:00 2023-06-24 23:00:00             FNG1   
442779 2023-06-24 22:00:00 2023-06-24 23:00:00             ERV5   
439972 2023-06-24 22:00:00 2023-06-24 23:00:00             ERV3   
437165 2023-06-24 22:00:00 2023-06-24 23:00:00             ERV2   
751214 2023-06-24 22:00:00 2023-06-24 23:00:00             WWD1   

                          Asset Name Asset Grouping      Volume  \
420203                  EGC1 Shepard           EGC1  649.203740   
468042                GN3 Genesee #3            GN3  461.520771   
541144              KH3 Keephills #3            KH3  434.641217   
637926              SCL1 Syncrude #1           SCL1  420.030548   
462428                GN1 Genesee #1            GN1  401.062047   
...                              ...            ...         ...   
451200              FNG1 Fort Nelson           FNG1    0.000000   
442779      ERV5 eReserve5 Hughenden           ERV5    0.000000   
439972    ERV3 eReserve3 Mercer Hill           ERV3    0.000000   
437165  ERV2 eReserve2 Buffalo Creek           ERV2    0.000000   
751214                 WWD1 Weldwood           WWD1    0.000000   

        Maximum Capability  System Capability       Fuel Type  \
420203               868.0              868.0             GAS   
468042               466.0              466.0       DUAL FUEL   
541144               463.0              463.0             GAS   
637926               510.0              510.0             GAS   
462428               400.0              400.0            COAL   
...                    ...                ...             ...   
451200                73.0               73.0             GAS   
442779                20.0               20.0  ENERGY STORAGE   
439972                20.0               20.0  ENERGY STORAGE   
437165                20.0               20.0  ENERGY STORAGE   
751214                50.0               10.0           OTHER   

          Sub Fuel Type  Planning Area     Region  Asset Utilization  
420203   COMBINED_CYCLE              6    Calgary           0.747931  
468042        DUAL FUEL             40   Edmonton           0.990388  
541144  GAS_FIRED_STEAM             40   Edmonton           0.938750  
637926     COGENERATION             25  Northeast           0.823589  
462428             COAL             40   Edmonton           1.002655  
...                 ...            ...        ...                ...  
451200   COMBINED_CYCLE             17  Northwest           0.000000  
442779   ENERGY STORAGE             37    Central           0.000000  
439972   ENERGY STORAGE             20  Northwest           0.000000  
437165   ENERGY STORAGE             32    Central           0.000000  
751214          Biomass             29    Central           0.000000  

[182 rows x 13 columns]

In [35]:
# June 26
june_26 = df_generation.loc[df_generation['Date (MST)'] == '2023-06-26 22:00']
june_26.sort_values('Volume', ascending = False)

Date (MST)          Date (MPT) Asset Short Name  \
758523 2023-06-26 22:00:00 2023-06-26 23:00:00             EGC1   
760701 2023-06-26 22:00:00 2023-06-26 23:00:00              GN3   
768203 2023-06-26 22:00:00 2023-06-26 23:00:00             SCL1   
763968 2023-06-26 22:00:00 2023-06-26 23:00:00              KH3   
760459 2023-06-26 22:00:00 2023-06-26 23:00:00              GN1   
...                    ...                 ...              ...   
760217 2023-06-26 22:00:00 2023-06-26 23:00:00             GEN5   
759975 2023-06-26 22:00:00 2023-06-26 23:00:00             FNG1   
759612 2023-06-26 22:00:00 2023-06-26 23:00:00             ERV5   
773188 2023-06-26 22:00:00 2023-06-26 23:00:00             WWD1   
770502 2023-06-26 22:00:00 2023-06-26 23:00:00             SUM1   

                      Asset Name Asset Grouping      Volume  \
758523              EGC1 Shepard           EGC1  720.965525   
760701            GN3 Genesee #3            GN3  463.426383   
768203          SCL1 Syncrude #1           SCL1  410.838297   
763968          KH3 Keephills #3            KH3  405.283714   
760459            GN1 Genesee #1            GN1  399.135938   
...                          ...            ...         ...   
760217         GEN5 Carson Creek           GEN5    0.000000   
759975          FNG1 Fort Nelson           FNG1    0.000000   
759612  ERV5 eReserve5 Hughenden           ERV5    0.000000   
773188             WWD1 Weldwood           WWD1    0.000000   
770502           SUM1 Summerview           SUM1   -0.001778   

        Maximum Capability  System Capability       Fuel Type  \
758523               868.0              868.0             GAS   
760701               466.0              466.0       DUAL FUEL   
768203               510.0              510.0             GAS   
763968               463.0              463.0             GAS   
760459               400.0              400.0            COAL   
...                    ...                ...             ...   
760217                15.0               15.0             GAS   
759975                73.0               73.0             GAS   
759612                20.0               20.0  ENERGY STORAGE   
773188                50.0               10.0           OTHER   
770502                10.0               10.0  ENERGY STORAGE   

          Sub Fuel Type  Planning Area     Region  Asset Utilization  
758523   COMBINED_CYCLE              6    Calgary           0.830605  
760701        DUAL FUEL             40   Edmonton           0.994477  
768203     COGENERATION             25  Northeast           0.805565  
763968  GAS_FIRED_STEAM             40   Edmonton           0.875343  
760459             COAL             40   Edmonton           0.997840  
...                 ...            ...        ...                ...  
760217     SIMPLE_CYCLE             26  Northwest           0.000000  
759975   COMBINED_CYCLE             17  Northwest           0.000000  
759612   ENERGY STORAGE             37    Central           0.000000  
773188          Biomass             29    Central           0.000000  
770502   ENERGY STORAGE             53      South          -0.000178  

[182 rows x 13 columns]

In [36]:
# Remove the duplicated rows for June 25 at 22:00 by removing rows where the asset name shows up at the same time more than once
df_generation_2 = df_generation.drop_duplicates(subset=['Date (MST)', 'Asset Short Name'], keep = 'first').reset_index(drop = True)

# from the new dataframe, repeat the above test to see how many rows are returned
june_25_2 = df_generation_2.loc[df_generation_2['Date (MST)'] == '2023-06-25 22:00']
june_25_2.sort_values('Volume', ascending = False)

Date (MST)          Date (MPT) Asset Short Name  \
420227 2023-06-25 22:00:00 2023-06-25 23:00:00             EGC1   
468066 2023-06-25 22:00:00 2023-06-25 23:00:00              GN3   
541168 2023-06-25 22:00:00 2023-06-25 23:00:00              KH3   
637950 2023-06-25 22:00:00 2023-06-25 23:00:00             SCL1   
462452 2023-06-25 22:00:00 2023-06-25 23:00:00              GN1   
...                    ...                 ...              ...   
451224 2023-06-25 22:00:00 2023-06-25 23:00:00             FNG1   
439996 2023-06-25 22:00:00 2023-06-25 23:00:00             ERV3   
437189 2023-06-25 22:00:00 2023-06-25 23:00:00             ERV2   
434382 2023-06-25 22:00:00 2023-06-25 23:00:00             ERV1   
751214 2023-06-25 22:00:00 2023-06-25 23:00:00             WWD1   

                          Asset Name Asset Grouping      Volume  \
420227                  EGC1 Shepard           EGC1  640.425430   
468066                GN3 Genesee #3            GN3  449.374494   
541168              KH3 Keephills #3            KH3  421.554772   
637950              SCL1 Syncrude #1           SCL1  415.500345   
462452                GN1 Genesee #1            GN1  402.335326   
...                              ...            ...         ...   
451224              FNG1 Fort Nelson           FNG1    0.000000   
439996    ERV3 eReserve3 Mercer Hill           ERV3    0.000000   
437189  ERV2 eReserve2 Buffalo Creek           ERV2    0.000000   
434382        ERV1 eReserve1 Rycroft           ERV1    0.000000   
751214                 WWD1 Weldwood           WWD1    0.000000   

        Maximum Capability  System Capability       Fuel Type  \
420227               868.0              868.0             GAS   
468066               466.0              466.0       DUAL FUEL   
541168               463.0              463.0             GAS   
637950               510.0              510.0             GAS   
462452               400.0              400.0            COAL   
...                    ...                ...             ...   
451224                73.0               73.0             GAS   
439996                20.0               20.0  ENERGY STORAGE   
437189                20.0               20.0  ENERGY STORAGE   
434382                20.0               20.0  ENERGY STORAGE   
751214                50.0               10.0           OTHER   

          Sub Fuel Type  Planning Area     Region  Asset Utilization  
420227   COMBINED_CYCLE              6    Calgary           0.737817  
468066        DUAL FUEL             40   Edmonton           0.964323  
541168  GAS_FIRED_STEAM             40   Edmonton           0.910485  
637950     COGENERATION             25  Northeast           0.814707  
462452             COAL             40   Edmonton           1.005838  
...                 ...            ...        ...                ...  
451224   COMBINED_CYCLE             17  Northwest           0.000000  
439996   ENERGY STORAGE             20  Northwest           0.000000  
437189   ENERGY STORAGE             32    Central           0.000000  
434382   ENERGY STORAGE             20  Northwest           0.000000  
751214          Biomass             29    Central           0.000000  

[182 rows x 13 columns]

In [37]:
# Repeat pivot table process with new dataframe to confirm outliers are removed
hourly_volume_df_2 = pd.pivot_table(df_generation_2, index = ['Date (MST)'], values = 'Volume', columns=['Fuel Type'], aggfunc='sum')
hourly_volume_df_2['Total Volume (MW)'] = hourly_volume_df_2[fuel_types].sum(axis=1)
hourly_volume_df_2.sort_values('Total Volume (MW)', ascending=False)

Fuel Type                  COAL   DUAL FUEL  ENERGY STORAGE          GAS  \
Date (MST)                                                                 
2023-01-23 10:00:00  817.470098  465.572827       45.741942  7943.036332   
2023-01-04 11:00:00  814.401373  463.833160        0.000000  7677.549624   
2023-02-24 14:00:00  814.341375  463.992550       -5.072503  7884.296063   
2023-02-24 15:00:00  815.415099  463.379159       -3.862059  7773.999193   
2023-01-06 11:00:00  814.999049  464.255827       -0.145445  7474.224570   
...                         ...         ...             ...          ...   
2023-05-07 01:00:00  302.724660  465.472104       -0.990444  5677.246546   
2023-05-28 05:00:00  803.744156  464.873881       -0.995889  5252.110418   
2023-05-28 04:00:00  803.614267  464.904773       -2.725722  5331.855850   
2023-05-07 02:00:00  298.398160  465.296827        0.000000  5629.051228   
2023-05-07 03:00:00  296.866327  465.299994        0.000000  5658.824619   

Fuel Type                 HYDRO       OTHER       SOLAR         WIND  \
Date (MST)                                                             
2023-01-23 10:00:00  140.702962  271.054301   51.532166  1923.733835   
2023-01-04 11:00:00  162.139227  269.446785  567.215049  1695.409923   
2023-02-24 14:00:00   87.971885  261.618407  628.555724  1509.295260   
2023-02-24 15:00:00  106.894894  265.549528  649.920766  1542.132724   
2023-01-06 11:00:00  135.156134  277.096481  504.583993  1928.993158   
...                         ...         ...         ...          ...   
2023-05-07 01:00:00  156.819058  209.474652    0.000000   621.639687   
2023-05-28 05:00:00  182.629757  256.220025  102.951032   361.649010   
2023-05-28 04:00:00  201.649363  254.425480    5.092000   356.830461   
2023-05-07 02:00:00  175.325419  208.868525    0.000000   632.341812   
2023-05-07 03:00:00  198.523500  209.025106    0.000000   567.860398   

Fuel Type            Total Volume (MW)  
Date (MST)                              
2023-01-23 10:00:00       11658.844463  
2023-01-04 11:00:00       11649.995142  
2023-02-24 14:00:00       11644.998762  
2023-02-24 15:00:00       11613.429304  
2023-01-06 11:00:00       11599.163767  
...                                ...  
2023-05-07 01:00:00        7432.386263  
2023-05-28 05:00:00        7423.182391  
2023-05-28 04:00:00        7415.646471  
2023-05-07 02:00:00        7409.281971  
2023-05-07 03:00:00        7396.399944  

[4343 rows x 9 columns]

#### Removing the duplicated rows appears to have worked; continue EDA analysis using this dataframe.

In [38]:
hourly_volume_df_2.describe()

Fuel Type         COAL    DUAL FUEL  ENERGY STORAGE          GAS        HYDRO  \
count      4343.000000  4343.000000     4343.000000  4343.000000  4343.000000   
mean        746.367270   459.965954       -0.424650  6726.520055   201.715021   
std         124.481197    32.410383        5.460892   893.930002   123.785300   
min         261.151496     0.000000      -25.269610  4323.263267    55.820617   
25%         762.413960   464.612160       -1.053194  6077.998118   124.073841   
50%         798.660880   465.184512        0.000000  6775.545045   161.200399   
75%         815.077238   465.694577        0.000000  7418.530253   230.429209   
max         821.208544   468.579993       74.199997  8811.276138   690.780322   

Fuel Type        OTHER        SOLAR         WIND  Total Volume (MW)  
count      4343.000000  4343.000000  4343.000000        4343.000000  
mean        252.424441   276.865369  1036.820594        9700.254054  
std          37.940479   364.363541   714.410373         899.522262  
min         127.428406    -0.244513     3.305500        7396.399944  
25%         233.495164     0.000000   392.362842        9051.031048  
50%         260.240773    15.900611   939.408382        9811.268736  
75%         278.075173   547.295895  1629.894398       10379.735340  
max         336.669533  1127.707101  2824.874941       11658.844463

#### From the above table, we can extract the following information:

1. The average hourly total volume production is 9.7 GW
2. On average, Natural Gas plants account for 69.3% of energy production each hour
3. At their maximum combined volumes (assuming the date and time are the same), solar and wind could only theoretically currently generate 40% of the average power required from the grid

In [39]:
# save assets dataset for visualizing in Tableau
#assets_per_region.to_csv('assets.csv')

In [40]:
# save changes made to df_generation for visualizing in Tableau
#df_generation.to_csv(''CSD Generation (Hourly) - 2023-01 to 2023-06_1.csv')

# 2.0 - Second Dataset - Hourly Metered Volumes and Pool Prices from 01-Jan-2010 to 08-Aug-2023

## 2.1 - EDA and Cleaning

In [41]:
# load second dataset which shows the Total Net Generation (MW) produced by each energy asset in Alberta
df_price = pd.read_csv('Hourly_Metered_Volumes_and_Pool_Price_and_AIL.csv')
df_price.shape

(119040, 213)

In [42]:
# display all columns
pd.set_option('display.max_columns', None)
df_price.head()

Date_Begin_GMT Date_Begin_Local    AFG1  AKE1  ALP1  ALP2  APS1     BCR2  \
0   1/1/2010 7:00    1/1/2010 0:00  5.2305   0.0   0.0   0.0   0.0  11.8714   
1   1/1/2010 8:00    1/1/2010 1:00  5.2360   0.0   0.0   0.0   0.0  11.8580   
2   1/1/2010 9:00    1/1/2010 2:00  5.5605   0.0   0.0   0.0   0.0  11.8566   
3  1/1/2010 10:00    1/1/2010 3:00  5.6250   0.0   0.0   0.0   0.0  11.8467   
4  1/1/2010 11:00    1/1/2010 4:00  7.1375   0.0   0.0   0.0   0.0  11.8596   

   BCRK        BIG       BOW1       BR3       BR4       BR5  BRA    BTR1  \
0   0.0  26.573744  42.754208  149.1570  156.2886  383.1201  0.0  0.0000   
1   0.0  26.753066  42.066657  149.7589  155.4531  382.9625  0.0  0.0000   
2   0.0  26.544753  45.459836  149.2844  151.4950  376.4841  0.0  0.0000   
3   0.0  26.826099  38.234323  149.4653  151.2895  374.1055  0.0  0.0000   
4   0.0  26.861537  44.831311  150.0822  151.4089  374.8381  0.0  0.1113   

   CAL1  CHIN  CMH1  CNR5  CR1  CRS1  CRS2  CRS3  CRW1   DAI1    DKSN  \
0   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0  2.394  4.7568   
1   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0  2.856  4.7568   
2   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0  2.226  4.7568   
3   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0  2.072  4.7544   
4   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0  2.226  4.7688   

         DOWG  DRW1     DV1       EAGL     EC01    EC04  ENC1  ENC2  ENC3  \
0  126.860491   0.0  7.1244  24.437387  40.6058  19.116   0.0   0.0   0.0   
1  118.038190   0.0  6.0588  23.965469  25.0991  18.800   0.0   0.0   0.0   
2   84.194438   0.0  7.2900  24.471967   8.5649  18.716   0.0   0.0   0.0   
3   80.459613   0.0  7.5420  24.379821   1.2963  17.760   0.0   0.0   0.0   
4   80.432869   0.0  7.2828  24.599399   1.2996  16.876   0.0   0.0   0.0   

       FNG1         GN1         GN2       GN3    GOC1     GPEC  GWW1  \
0  13.05024  390.362194  388.703406  450.0084  2.9862  14.4051   0.0   
1  13.27632  391.990632  388.673368  450.3962  2.9070  14.4389   0.0   
2  13.37832  392.053344  388.735456  450.0711  2.9610  14.7085   0.0   
3  13.61136  392.086054  388.214746  450.0501  3.1608  15.1453   0.0   
4  13.85232  392.547075  388.676925  450.2333  3.1950  14.6274   0.0   

          HRM  ICP1  IEW1     IOR1      JOF1        KH1       KH2  KHW1  ME02  \
0  125.467731   0.0   0.0  52.5424  55.63535  346.22508  0.870919   0.0   0.0   
1  125.404366   0.0   0.0  52.8890   6.98300  354.95541  0.000000   0.0   0.0   
2  126.500882   0.0   0.0  52.3071   5.26648  356.76656  0.000000   0.0   0.0   
3  126.244356   0.0   0.0  52.3084   0.71940  357.67663  0.000000   0.0   0.0   
4  124.792731   0.0   0.0  52.1768   2.26765  363.45636  0.000000   0.0   0.0   

   ME03  ME04    MEG1     MKR1      MKRC  NPC1  NPP1     NX01    NX02    OMRH  \
0   0.0   0.0  84.096  50.3675  170.5408   0.0   0.0  76.3530  23.588  3.3606   
1   0.0   0.0  84.304  52.5751  170.4663   0.0   0.0  55.4279  21.658  3.3606   
2   0.0   0.0  83.984  49.3939  170.4563   0.0   0.0  58.6236  26.378  3.3366   
3   0.0   0.0  84.240  46.6872  170.2688   0.0   0.0  55.1697  27.536  3.3606   
4   0.0   0.0  84.056  51.5074  171.4955   0.0   0.0  68.2187  26.052  3.3606   

   PH1      PR1    PW01  RB1  RB2  RB3     RB5  RG10  RG8  RG9     RL1  RYMD  \
0  0.0  15.7386  5.5143  0.0  0.0  0.0  13.904   0.0  0.0  0.0  43.540   0.0   
1  0.0  15.9696  5.5105  0.0  0.0  0.0  12.328   0.0  0.0  0.0  43.610   0.0   
2  0.0  15.5492  5.5211  0.0  0.0  0.0   8.684   0.0  0.0  0.0  43.582   0.0   
3  0.0  15.3975  5.5060  0.0  0.0  0.0   8.764   0.0  0.0  0.0  43.610   0.0   
4  0.0  15.2048  5.5242  0.0  0.0  0.0   8.864   0.0  0.0  0.0  43.624   0.0   

     SCL1     SCR1    SCR2     SCR3        SD1        SD2        SD3  \
0  45.296  188.216   2.932  10.9994  258.17587  260.49263  313.53413   
1  35.872  181.352   7.354  13.9316  260.65199  260.63678  312.28792   
2  37.328  165.088  24.438  20.6898  257.58459  253.45023  313.54

In [43]:
# remove columns which are not needed for this analysis
df_price = df_price.drop(['Date_Begin_GMT', 'DAY_AHEAD_POOL_PRICE'], axis=1)

#### NaN Values in this dataset represent assets which are now operational that were not operational 01-Jan-2010. For the purpose of this analysis, I will fill these values with zeros.

In [44]:
df_price = df_price.fillna(0)
df_price.head()

Date_Begin_Local    AFG1  AKE1  ALP1  ALP2  APS1     BCR2  BCRK        BIG  \
0    1/1/2010 0:00  5.2305   0.0   0.0   0.0   0.0  11.8714   0.0  26.573744   
1    1/1/2010 1:00  5.2360   0.0   0.0   0.0   0.0  11.8580   0.0  26.753066   
2    1/1/2010 2:00  5.5605   0.0   0.0   0.0   0.0  11.8566   0.0  26.544753   
3    1/1/2010 3:00  5.6250   0.0   0.0   0.0   0.0  11.8467   0.0  26.826099   
4    1/1/2010 4:00  7.1375   0.0   0.0   0.0   0.0  11.8596   0.0  26.861537   

        BOW1       BR3       BR4       BR5  BRA    BTR1  CAL1  CHIN  CMH1  \
0  42.754208  149.1570  156.2886  383.1201  0.0  0.0000   0.0   0.0   0.0   
1  42.066657  149.7589  155.4531  382.9625  0.0  0.0000   0.0   0.0   0.0   
2  45.459836  149.2844  151.4950  376.4841  0.0  0.0000   0.0   0.0   0.0   
3  38.234323  149.4653  151.2895  374.1055  0.0  0.0000   0.0   0.0   0.0   
4  44.831311  150.0822  151.4089  374.8381  0.0  0.1113   0.0   0.0   0.0   

   CNR5  CR1  CRS1  CRS2  CRS3  CRW1   DAI1    DKSN        DOWG  DRW1     DV1  \
0   0.0  0.0   0.0   0.0   0.0   0.0  2.394  4.7568  126.860491   0.0  7.1244   
1   0.0  0.0   0.0   0.0   0.0   0.0  2.856  4.7568  118.038190   0.0  6.0588   
2   0.0  0.0   0.0   0.0   0.0   0.0  2.226  4.7568   84.194438   0.0  7.2900   
3   0.0  0.0   0.0   0.0   0.0   0.0  2.072  4.7544   80.459613   0.0  7.5420   
4   0.0  0.0   0.0   0.0   0.0   0.0  2.226  4.7688   80.432869   0.0  7.2828   

        EAGL     EC01    EC04  ENC1  ENC2  ENC3      FNG1         GN1  \
0  24.437387  40.6058  19.116   0.0   0.0   0.0  13.05024  390.362194   
1  23.965469  25.0991  18.800   0.0   0.0   0.0  13.27632  391.990632   
2  24.471967   8.5649  18.716   0.0   0.0   0.0  13.37832  392.053344   
3  24.379821   1.2963  17.760   0.0   0.0   0.0  13.61136  392.086054   
4  24.599399   1.2996  16.876   0.0   0.0   0.0  13.85232  392.547075   

          GN2       GN3    GOC1     GPEC  GWW1         HRM  ICP1  IEW1  \
0  388.703406  450.0084  2.9862  14.4051   0.0  125.467731   0.0   0.0   
1  388.673368  450.3962  2.9070  14.4389   0.0  125.404366   0.0   0.0   
2  388.735456  450.0711  2.9610  14.7085   0.0  126.500882   0.0   0.0   
3  388.214746  450.0501  3.1608  15.1453   0.0  126.244356   0.0   0.0   
4  388.676925  450.2333  3.1950  14.6274   0.0  124.792731   0.0   0.0   

      IOR1      JOF1        KH1       KH2  KHW1  ME02  ME03  ME04    MEG1  \
0  52.5424  55.63535  346.22508  0.870919   0.0   0.0   0.0   0.0  84.096   
1  52.8890   6.98300  354.95541  0.000000   0.0   0.0   0.0   0.0  84.304   
2  52.3071   5.26648  356.76656  0.000000   0.0   0.0   0.0   0.0  83.984   
3  52.3084   0.71940  357.67663  0.000000   0.0   0.0   0.0   0.0  84.240   
4  52.1768   2.26765  363.45636  0.000000   0.0   0.0   0.0   0.0  84.056   

      MKR1      MKRC  NPC1  NPP1     NX01    NX02    OMRH  PH1      PR1  \
0  50.3675  170.5408   0.0   0.0  76.3530  23.588  3.3606  0.0  15.7386   
1  52.5751  170.4663   0.0   0.0  55.4279  21.658  3.3606  0.0  15.9696   
2  49.3939  170.4563   0.0   0.0  58.6236  26.378  3.3366  0.0  15.5492   
3  46.6872  170.2688   0.0   0.0  55.1697  27.536  3.3606  0.0  15.3975   
4  51.5074  171.4955   0.0   0.0  68.2187  26.052  3.3606  0.0  15.2048   

     PW01  RB1  RB2  RB3     RB5  RG10  RG8  RG9     RL1  RYMD    SCL1  \
0  5.5143  0.0  0.0  0.0  13.904   0.0  0.0  0.0  43.540   0.0  45.296   
1  5.5105  0.0  0.0  0.0  12.328   0.0  0.0  0.0  43.610   0.0  35.872   
2  5.5211  0.0  0.0  0.0   8.684   0.0  0.0  0.0  43.582   0.0  37.328   
3  5.5060  0.0  0.0  0.0   8.764   0.0  0.0  0.0  43.610   0.0  40.256   
4  5.5242  0.0  0.0  0.0   8.864   0.0  0.0  0.0  43.624   0.0  44.320   

      SCR1    SCR2     SCR3        SD1        SD2        SD3        SD4  \
0  188.216   2.932  10.9994  258.17587  260.49263  313.53413  356.57294   
1  181.352   7.354  13.9316  260.65199  260.63678  312.28792  355.43224   
2  165.088  24.438  20.6898  257.58459  253.45023  313.54758  336.47152   
3  145.424  21.036  27.6651 

In [63]:
# Define all assets
assets = df_price.columns[1:-18]

In [65]:
# List out what assets fall under what fuel type.
coal_list = ['GN1', 'GN2', 'RG10', 'RG8', 'RG9', 'SD1', 'SD2', 'SD3', 'WB4']

dual_fuel_list = ['GN3', 'SD4', 'SD5', 'SD6']

energy_storage_list = ['SUM1', 'ERV1', 'ERV2', 'ERV3', 'ERV5']

gas_list = ['ALP1', 'ALP2', 'APS1', 'BCR2', 'BCRK', 'BR3', 'BR4', 'BR5', 'CAL1', 'CMH1', 'CNR5', 'CRS1', 'CRS2', 'CRS3', 'DOWG', 'DRW1', 'EC01', 'EC04',
           'ENC1', 'ENC2', 'ENC3', 'FNG1', 'HRM', 'IOR1', 'JOF1', 'KH1', 'KH2', 'ME02', 'ME03', 'ME04', 'MEG1', 'MKR1', 'MKRC', 'NPC1', 'NPP1', 'NX01',
           'NX02', 'PH1', 'PR1', 'PW01', 'RB1', 'RB2', 'RB3', 'RB5', 'RL1', 'SCL1', 'SCR1', 'SH1', 'SH2', 'SHCG', 'ST1', 'ST2', 'WWD2', 'HMT1', 'KH3',
           'UOC1', 'GEN6', 'ANC1', 'EGC1', 'PEC1', 'WCD1', 'CRG1', 'HSM1', 'NAT1', 'SDH1', 'CL01', 'FH1', 'MUL1', 'BHL1', 'BFD1', 'COD1', 'PMB1', 'IOR4',
           'SET1', 'HRT1', 'NPC3', 'EPS1', 'HRV1', 'TRH1', 'HRV2', 'WLC1'] 

hydro_list = ['BIG', 'BOW1', 'BRA', 'CHIN', 'DKSN', 'ICP1', 'OMRH', 'RYMD', 'TAY1', 'TC01', 'TC02', 'TLM2', 'UOA1', 'VVW1', 'VVW2']

other_list = ['AFG1', 'DAI1', 'EAGL', 'GOC1', 'GPEC', 'WEY1', 'WST1', 'WWD1', 'CRR1', 'NRG3', 'SLP1', 'BON1', 'SRL1']

solar_list = ['BSC1', 'VXH1', 'HUL1', 'INF1', 'SUF1', 'CLR1', 'CLR2', 'BRD1', 'BUR1', 'WEF1', 'HYS1', 'JER1', 'TVS1', 'STR1', 'STR2', 'BRK1', 'BRK2', 'COL1',
             'NMK1', 'MON1', 'CRD1', 'CRD2', 'KKP1', 'KKP2', 'VNC1', 'CLY1', 'CLY2', 'MIC1', 'BLS1', 'CHP1', 'MCH1', 'STV1', 'EMP1', 'DFT1', 'YNG1']

wind_list = ['AKE1', 'BTR1', 'CR1', 'CRW1', 'GWW1', 'IEW1', 'KHW1', 'SCR2', 'SCR3', 'TAB1', 'TAY2', 'IEW2', 'ARD1', 'NEP1', 'SCR4', 'HAL1', 'BSR1', 'OWF1',
            'BUL1', 'BUL2', 'WHT1', 'CRR2', 'RIV1', 'WRW1', 'WHT2', 'RTL1', 'WHE1', 'FMG1', 'HHW1', 'CYP2', 'HLD1', 'CYP1', 'LAN1', 'JNR3', 'GDP1', 'GRZ1',
            'SWP1']

# create list of all assets in the price dataset
price_assets = coal_list + dual_fuel_list + gas_list + hydro_list + other_list + solar_list + wind_list + energy_storage_list
price_asset_count = len(coal_list) + len(dual_fuel_list) + len(gas_list) + len(hydro_list) + len(other_list) + len(solar_list) + len(wind_list) + len(energy_storage_list)

# check to see if there are assets in the price dataset that are not in the generation dataset
generation_assets_set = set(generation_assets)
price_assets_set = set(price_assets)
missing_assets_generation = price_assets_set - generation_assets_set
print(list(missing_assets_generation))

['RG9', 'RB2', 'ST2', 'WLC1', 'NPC1', 'SD1', 'SD4', 'TAY2', 'HSM1', 'RB1', 'ST1', 'VNC1', 'DFT1', 'SD2', 'WWD2', 'PW01', 'RB3', 'SD3', 'YNG1', 'SD5', 'CRW1', 'WB4', 'KH1', 'RG10', 'RG8', 'BR3']


#### Assets in the price dataset that are not in generation dataset (retired assets or converted assets)
**Total MC missing = 1,993 MW**
- Coal:
    - RG10 - Unknown MW
    - RG8 - Unknown MW
    - RG9 - Unknown MW
    - SD1 - 280
    - SD2 - 280
    - SD3 - 368
    - WB4 - Unknown MW
- Dual Fuel:
    - SD4 - 406 (coal to gas conversion)
    - SD5 - 406 (coal to gas conversion)  
- Gas:
    - BR3 - 149 MW
    - NPC1 - 11 MW
    - PW01 - 5 MW
    - KH1 - Unknown MW
    - RB1 - Unknown MW
    - RB2 - Unknown MW
    - RB3 - Unknown MW
    - ST1 - Unknown MW
    - ST2 - Unknown MW
    - WWD2 - 30 MW
    - HSM1 - 6 MW
    - WLC1 - 9 MW
- Solar:    
    - DFT1 - 37 MW
    - YNG1 - 6 MW
- Wind:
    - CRW1 - Unknown MW
    - TAY2 - Unknown MW
    - VNC1 - Unknown MW
- Unknown:
    - DV1 - Unknown MW
    - MFG1 - Unknown MW

In [66]:
missing_assets_price = generation_assets_set - price_assets_set
print(list(missing_assets_price))

['ALS1', 'IOR2', 'CCMH', 'SCR5', 'NPC2', 'CRE3', 'SCR6', 'IOR3', 'GEN5', 'VCN1']


#### Assets in the generation dataset that are missing from the price dataset
- **Total MC Missing =  1,270 MW**
- Gas:
  - ALS1 - 106 MW
  - IOR2 - 195 MW
  - IOR3 - 84 MW
  - GEN5 - 15 MW
  - NPC2 - 9 MW
  - SCR5 - 376 MW
  - SCR6 - 401 MW
- Other:
  - CCMH - 42 MW
- Solar: 
  - VCN1 - 22 MW
- Wind: 
  - CRE3 - 20 MW 

#### Create new dataframe that sums all the assets by fuel type

In [67]:
df_price_3 = df_price.copy()
# add new fuel type totals
df_price_3['Total Coal Volume (MW)'] = df_price_3[coal_list].sum(axis=1)
df_price_3['Total Dual Fuel Volume (MW)'] = df_price_3[dual_fuel_list].sum(axis=1)
df_price_3['Total Energy Storage Volume (MW)'] = df_price_3[energy_storage_list].sum(axis=1)
df_price_3['Total Gas Volume (MW)'] = df_price_3[gas_list].sum(axis=1)
df_price_3['Total Hydro Volume (MW)'] = df_price_3[hydro_list].sum(axis=1)
df_price_3['Total Other Volume (MW)'] = df_price_3[other_list].sum(axis=1)
df_price_3['Total Solar Volume (MW)'] = df_price_3[solar_list].sum(axis=1)
df_price_3['Total Wind Volume (MW)'] = df_price_3[wind_list].sum(axis=1)
df_price_3['Total Volume (MW)'] = df_price_3[['Total Coal Volume (MW)', 'Total Dual Fuel Volume (MW)', 'Total Energy Storage Volume (MW)', 'Total Gas Volume (MW)',
            'Total Hydro Volume (MW)', 'Total Other Volume (MW)', 'Total Solar Volume (MW)', 'Total Wind Volume (MW)']].sum(axis=1)
# add new column summing up imports and export values (Net Actual Interchange)
df_price_3['Net Actual Interchange (MW)'] = df_price_3[['EXPORT_BC', 'EXPORT_MT', 'EXPORT_SK']].sum(axis=1) - df_price_3[['IMPORT_BC', 'IMPORT_MT', 'IMPORT_SK']].sum(axis=1)

In [68]:
df_price_3

Date_Begin_Local     AFG1     AKE1    ALP1    ALP2  APS1       BCR2  \
0         1/1/2010 0:00   5.2305   0.0000  0.0000  0.0000   0.0  11.871400   
1         1/1/2010 1:00   5.2360   0.0000  0.0000  0.0000   0.0  11.858000   
2         1/1/2010 2:00   5.5605   0.0000  0.0000  0.0000   0.0  11.856600   
3         1/1/2010 3:00   5.6250   0.0000  0.0000  0.0000   0.0  11.846700   
4         1/1/2010 4:00   7.1375   0.0000  0.0000  0.0000   0.0  11.859600   
...                 ...      ...      ...     ...     ...   ...        ...   
119035  7/31/2023 20:00  18.2925  11.8928  5.1280  6.6640   0.0  31.217982   
119036  7/31/2023 21:00  18.0205  14.9536  6.1680  6.6880   0.0  31.203876   
119037  7/31/2023 22:00  18.2090  18.5623  6.3360  6.6960   0.0  31.248870   
119038  7/31/2023 23:00  18.0685  19.2688  6.3360  6.6960   0.0  31.232824   
119039    8/1/2023 0:00  17.7430  30.3356  0.9832  0.9289   0.0  31.195405   

             BCRK        BIG       BOW1       BR3       BR4         BR5  \
0        0.000000  26.573744  42.754208  149.1570  156.2886  383.120100   
1        0.000000  26.753066  42.066657  149.7589  155.4531  382.962500   
2        0.000000  26.544753  45.459836  149.2844  151.4950  376.484100   
3        0.000000  26.826099  38.234323  149.4653  151.2895  374.105500   
4        0.000000  26.861537  44.831311  150.0822  151.4089  374.838100   
...           ...        ...        ...       ...       ...         ...   
119035  52.710018  85.017730  59.633204    0.0000    0.0000  176.687317   
119036  52.996124  35.086668  51.480938    0.0000    0.0000  176.707894   
119037  54.015130  33.476498  51.096185    0.0000    0.0000  179.629381   
119038  54.355176  30.797040  37.323485    0.0000    0.0000  174.933010   
119039  54.776595  45.922064  39.119698    0.0000    0.0000  177.030716   

              BRA     BTR1      CAL1    CHIN      CMH1  CNR5     CR1  \
0        0.000000   0.0000    0.0000   0.000    0.0000   0.0  0.0000   
1        0.000000   0.0000    0.0000   0.000    0.0000   0.0  0.0000   
2        0.000000   0.0000    0.0000   0.000    0.0000   0.0  0.0000   
3        0.000000   0.0000    0.0000   0.000    0.0000   0.0  0.0000   
4        0.000000   0.1113    0.0000   0.000    0.0000   0.0  0.0000   
...           ...      ...       ...     ...       ...   ...     ...   
119035  23.251320  10.7360  256.6116  11.376  105.1604   0.0  4.9590   
119036  20.431939  10.1120  251.2934  11.360  117.0903   0.0  1.5526   
119037  26.311160   8.0240  252.6095  11.360   68.1447   0.0  1.4174   
119038   0.000000  21.4400  254.7035  11.376   82.8578   0.0  5.4442   
119039   0.000000  24.4080  205.2834  11.360   59.5019   0.0  2.9948   

             CRS1       CRS2  CRS3   CRW1     DAI1     DKSN        DOWG  DRW1  \
0        0.000000   0.000000   0.0  0.000   2.3940   4.7568  126.860491   0.0   
1        0.000000   0.000000   0.0  0.000   2.8560   4.7568  118.038190   0.0   
2        0.000000   0.000000   0.0  0.000   2.2260   4.7568   84.194438   0.0   
3        0.000000   0.000000   0.0  0.000   2.0720   4.7544   80.459613   0.0   
4        0.000000   0.000000   0.0  0.000   2.2260   4.7688   80.432869   0.0   
...           ...        ...   ...    ...      ...      ...         ...   ...   
119035  39.508907  38.957733   0.0  6.200  20.1488  11.3699  147.281550   0.0   
119036  15.891784   2.933670   0.0  4.408  20.7186  11.3718  144.355273   0.0   
119037   0.000000   0.000000   0.0  3.184  10.5910  11.3714  142.497766   0.0   
119038   0.000000   0.000000   0.0  1.736  12.9570  11.3696  143.069773   0.0   
119039   0.000000   0.000000   0.0  2.440   7.2016  11.3671  148.027756   0.0   

           DV1       EAGL     EC01    EC04       ENC1       ENC2       ENC3  \
0       7.1244  24.437387  40.6058  19.116   0.000000   0.000000   0.000000   
1       6.0588  23.965469  25.0991  18.800   0.000000   0.000000   0.000000   
2       7.2900  24.471967   8.5649  18.716   0.000000   0.000000   0.000000   
3       7.5420 

In [69]:
# remove individual asset columns
df_price_3a = df_price_3.iloc[:,:1]
df_price_3b = df_price_3.iloc[:,-19:]
df_price_3 = df_price_3a.join(df_price_3b)
# sum asset volumes to confirm calculated volume = total volume
df_price_3['Calculated Generated Volume'] = df_price[assets].sum(axis=1)
# reorganize columns
df_price_3 = df_price_3[['Date_Begin_Local', 'Total Coal Volume (MW)', 'Total Dual Fuel Volume (MW)', 'Total Energy Storage Volume (MW)', 'Total Gas Volume (MW)', 
'Total Hydro Volume (MW)', 'Total Other Volume (MW)', 'Total Solar Volume (MW)', 'Total Wind Volume (MW)', 'EXPORT_BC', 'EXPORT_MT', 'EXPORT_SK', 'IMPORT_BC', 
'IMPORT_MT', 'IMPORT_SK', 'Net Actual Interchange (MW)', 'Total Volume (MW)', 'Calculated Generated Volume', 'ACTUAL_AIL', 'ACTUAL_POOL_PRICE']]
df_price_3

Date_Begin_Local  Total Coal Volume (MW)  Total Dual Fuel Volume (MW)  \
0         1/1/2010 0:00              1861.42200                   1538.81954   
1         1/1/2010 1:00              1865.16100                   1532.95787   
2         1/1/2010 2:00              1856.02633                   1459.49709   
3         1/1/2010 3:00              1835.94725                   1483.37852   
4         1/1/2010 4:00              1832.14724                   1480.59632   
...                 ...                     ...                          ...   
119035  7/31/2023 20:00               795.39200                    748.16826   
119036  7/31/2023 21:00               794.67200                    749.41869   
119037  7/31/2023 22:00               793.79200                    748.98194   
119038  7/31/2023 23:00               794.25600                    620.16036   
119039    8/1/2023 0:00               793.50400                    623.28127   

        Total Energy Storage Volume (MW)  Total Gas Volume (MW)  \
0                                 0.0000            2977.454511   
1                                 0.0000            2874.329086   
2                                 0.0000            2799.468980   
3                                 0.0000            2759.900759   
4                                 0.0000            2725.678230   
...                                  ...                    ...   
119035                            0.0776            4610.879205   
119036                            0.0780            4441.835571   
119037                            0.0780            4330.806266   
119038                            0.0777            4066.893622   
119039                            0.0735            3813.440435   

        Total Hydro Volume (MW)  Total Other Volume (MW)  \
0                    171.547252                49.453187   
1                    171.552623                49.403369   
2                    174.645489                49.927967   
3                    167.827122                50.382921   
4                    174.282248                51.785299   
...                         ...                      ...   
119035               400.795417               120.684188   
119036               343.000815               119.449576   
119037               309.774938               101.809038   
119038               266.444986               101.454302   
119039               278.589318               101.416272   

        Total Solar Volume (MW)  Total Wind Volume (MW)  EXPORT_BC  EXPORT_MT  \
0                      0.000000               36.252801        125        0.0   
1                      0.000000               57.551401        125        0.0   
2                      0.000000               76.991201        150        0.0   
3                      0.000000               83.687901        150        0.0   
4                      0.000000               97.121801        100        0.0   
...                         ...                     ...        ...        ...   
119035               109.373551              514.980067          0       26.0   
119036                 1.428123              708.449112          0       27.0   
119037                 0.001800              978.107312          0       27.0   
119038                 0.001900             1150.020631         40       26.0   
119039                 0.001700             1212.304228        100       26.0   

        EXPORT_SK  IMPORT_BC  IMPORT_MT  IMPORT_SK  \
0               0         50        0.0         44   
1               0          0        0.0         45   
2               0         50        0.0         45   
3               0         50        0.0         44   
4               0          0        0.0         95   
...           ...        ...        ...        ...   
119035         90        427        0.0          0   
119036         90        445        0.0          0   
119037         90        225        0.0          0   
119038 

In [70]:
# convert date column to datetime
df_price_3['Date_Begin_Local'] = pd.to_datetime(df_price_3['Date_Begin_Local'])

#### Export the df_price_3 dataset to csv for visualization in Tableau to show how the grid has changed since 2010

In [71]:
#df_price_3.to_csv('df_price.csv')